packages/discord-request/src/request.ts

```diff
@@ -0,0 +1,79 @@
+import type { ErrorBody } from "discord-error";
+import { DiscordError, isDiscordError } from "discord-error";
+import type { Client } from "./client.js";
+import { RateLimitError } from "./errors/rate-limit-error.js";
+import { RequestError } from "./errors/request-error.js";
+import { parse } from "./response.js";
+
+export async function request(
+	client: Client,
+	path: string,
+	init: RequestInit,
+	abortSignal: AbortSignal | null | undefined
+): Promise<ArrayBuffer | JSON | Response> {
+	// Setup the timeout signal
+	const signal = new AbortController();
+	const abortRequest = () => {
+		signal.abort();
+		clearAbort();
+	};
+
+	const abortTimeout = setTimeout(abortRequest, client.timeout);
+	const clearAbort = () => {
+		clearTimeout(abortTimeout);
+	};
+
+	abortSignal?.addEventListener("abort", abortRequest);
+
+	// Callbacks
+	client.onRequest?.(path, init);
+
+	// Send the request
+	const request = new Request(path, init);
+	let response: Response;
+
+	// Failed requests should bubble up
+	try {
+		response = await fetch(request, { signal: signal.signal });
+	} finally {
+		abortSignal?.removeEventListener("abort", abortRequest);
+		clearAbort();
+	}
+
+	// Successful Requests
+	if (response.ok) {
+		return parse(response);
+	}
+
+	// Rate Limited Requests
+	if (response.status === 429) {
+		throw new RateLimitError(response);
+	}
+
+	// If given a server error, retry the request
+	if (response.status >= 500 && response.status < 600) {
+		throw new RequestError(
+			path,
+			init,
+			// `response` has not yet been consumed by this point, it is safe to pass an uncloned version
+			response
```

Comment:
So ummm, this implies the consumer of the error is responsible for consuming the response within the error. That seems weird to me

Reply:
In theory most devs won't be looking at this error value, but if they do I think it's better to let them utilize the raw response as they see fit

Reply:
Yes, I know that most devs won't even use it. That's why it matters here. You are supposed to consume the body, and t his leaves it to the consumer.